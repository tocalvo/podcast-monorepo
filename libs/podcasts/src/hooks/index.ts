export * from './useItunesPodcasts';
export * from './useItunesPodcastDetail';
