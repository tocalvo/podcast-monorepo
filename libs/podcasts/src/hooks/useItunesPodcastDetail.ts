import { useQuery } from '@tanstack/react-query';
import {
  ITunesPodcastDetailQuery,
  ITunesPodcastDetailQueryExtended,
  ITunesPodcastKind,
} from '../models';
import { formatMillisecondsToMinSec } from '../utils';
import { processAllOriginsResponse, wrapUrl } from '../utils/url-prefixer';
import DOMPurify from 'isomorphic-dompurify';
import { useMemo } from 'react';

export const ITUNES_PODCAST_DETAIL_QUERY = 'ITUNES_PODCAST_DETAIL_QUERY';

export function useItunesPodcastDetail(id?: string) {
  const podcastsDetailQuery = useQuery<ITunesPodcastDetailQueryExtended>({
    queryKey: [ITUNES_PODCAST_DETAIL_QUERY, id],
    queryFn: async () => {
      const res = await fetch(
        wrapUrl(
          `https://itunes.apple.com/lookup?id=${id}&media=podcast&entity=podcastEpisode&limite=20`
        )
      );
      const parsed: ITunesPodcastDetailQuery = await processAllOriginsResponse(
        res
      );
      return {
        ...parsed,
        results: parsed.results.map((result) => ({
          ...result,
          sanitizedDescription: DOMPurify.sanitize(result.description || ''),
          releaseDateLocaleString: new Date(
            result.releaseDate
          ).toLocaleDateString(),
          trackTimeMinsSegs: formatMillisecondsToMinSec(result.trackTimeMillis),
        })),
      };
    },
    gcTime: 1000 * 60 * 60 * 24, // 24 hours cache
    staleTime: 1000 * 60 * 60 * 24, // 24 hours cache
    refetchOnWindowFocus: false,
  });

  const episodes = useMemo(
    () =>
      podcastsDetailQuery.data?.results.filter(
        (podcast) => podcast.kind === ITunesPodcastKind['podcast-episode']
      ) || [],

    [podcastsDetailQuery.data]
  );

  const findEpisodeById = (id?: string) =>
    episodes.find((episode) => episode.trackId === parseInt(id || ''));

  return {
    podcastDetail: podcastsDetailQuery.data,
    episodes,
    findEpisodeById,
  };
}

export default useItunesPodcastDetail;
