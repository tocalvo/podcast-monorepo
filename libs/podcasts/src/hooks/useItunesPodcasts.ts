import { useQuery } from '@tanstack/react-query';
import { ITunesPodcasts } from '../models';

import { processAllOriginsResponse, wrapUrl } from '../utils/url-prefixer';

export const ITUNES_PODCASTS_QUERY = 'ITUNES_PODCASTS_QUERY';

export function useItunesPodcasts() {
  const podcastsQuery = useQuery<ITunesPodcasts>({
    queryKey: [ITUNES_PODCASTS_QUERY],
    queryFn: async () => {
      const res = await fetch(
        wrapUrl(
          'https://itunes.apple.com/us/rss/toppodcasts/limit=100/genre=1310/json'
        )
      );
      return processAllOriginsResponse(res);
    },
    gcTime: 1000 * 60 * 60 * 24, // 24 hours
    staleTime: 1000 * 60 * 60 * 24, // 24 hours
    refetchOnWindowFocus: false,
  });

  const getPodcastResumeById = (id?: string) => {
    return podcastsQuery.data?.feed.entry.find(
      (podcast) => podcast.id.attributes['im:id'] === id
    );
  };

  return {
    getPodcastResumeById,
    podcasts: podcastsQuery.data?.feed.entry || [],
  };
}
