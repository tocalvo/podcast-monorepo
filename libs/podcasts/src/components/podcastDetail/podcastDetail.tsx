import { useItunesPodcastDetail, useItunesPodcasts } from '../../hooks';
import styles from './podcastDetail.module.css';
import PodcastBar from '../podcastBar/podcastBar';
import { Link, useParams } from 'react-router';

export const PodcastDetail: React.FC = () => {
  const { podcastId } = useParams();

  const { getPodcastResumeById } = useItunesPodcasts();
  const itunesPodcastResume = getPodcastResumeById(podcastId);

  const { episodes } = useItunesPodcastDetail(podcastId);

  if (!itunesPodcastResume) {
    return; // redirect to home or to 404
  }

  return (
    <div className={styles['container']}>
      <PodcastBar podcast={itunesPodcastResume} />
      <div className={styles['episodes-container']}>
        <div className={styles['card']}>
          <p>
            <b>Episodes: {episodes?.length}</b>
          </p>
        </div>
        <div className={`${styles['card']} ${styles['striped-table']}`}>
          <table>
            <thead>
              <tr>
                <th>Title</th>
                <th>Date</th>
                <th>Duration</th>
              </tr>
            </thead>
            <tbody>
              {episodes.map((episode, i) => (
                <tr key={i}>
                  <td>
                    <Link
                      className={styles['link']}
                      to={`/podcast/${podcastId}/episode/${episode.trackId}`}
                    >
                      {episode.trackName}
                    </Link>
                  </td>
                  <td>{episode.releaseDateLocaleString}</td>
                  <td>{episode.trackTimeMinsSegs}</td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      </div>
    </div>
  );
};
export default PodcastDetail;
