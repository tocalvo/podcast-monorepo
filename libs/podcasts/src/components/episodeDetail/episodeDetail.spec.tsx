import { render, screen } from '@testing-library/react';

import { EpisodeDetail } from './EpisodeDetail';
import { podcastListMock } from '../../__mocks__/podcast.mock';
import { podcastDetailMock } from '../../__mocks__/podcastDetail.mock';
import { BrowserRouter } from 'react-router';
import { useItunesPodcastDetail } from '../../hooks';
import { Mock } from 'vitest';

vi.mock('../../hooks', () => ({
  useItunesPodcasts: vi.fn(() => ({
    data: podcastListMock,
    isLoading: false,
    error: null,
  })),
  useItunesPodcastDetail: vi.fn(() => ({
    data: podcastDetailMock,
    isLoading: false,
    error: null,
  })),
}));

vi.mock('react-router', async () => {
  const actual = await vi.importActual('react-router');
  return {
    ...actual,
    useParams: vi.fn(() => ({
      podcastId: podcastListMock.feed.entry[0].id.attributes['im:id'],
      episodeId: podcastDetailMock.results[1].trackId,
    })),
  };
});

describe('EpisodeDetail', () => {
  beforeEach(() => {
    vi.restoreAllMocks();
  });

  it('should render successfully', () => {
    const { baseElement, container } = render(
      <BrowserRouter>
        <EpisodeDetail />
      </BrowserRouter>
    );
    expect(baseElement).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('should render the podcasts bar', () => {
    render(
      <BrowserRouter>
        <EpisodeDetail />
      </BrowserRouter>
    );

    expect(screen.queryByTestId('podcasts_bar')).toBeTruthy();
  });

  it('should render the episode title', () => {
    render(
      <BrowserRouter>
        <EpisodeDetail />
      </BrowserRouter>
    );

    expect(screen.getByRole('heading', { level: 1 }).textContent).toBe(
      'Episode 778 | "Bottom Of The 9th"'
    );
  });

  it('should sanitize the content', () => {
    const { container } = render(
      <BrowserRouter>
        <EpisodeDetail />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
    expect(screen.queryByTestId('episode_content')?.innerHTML).toContain(
      '<img src="x">'
    );
  });

  it('should render the audio', () => {
    const { container } = render(
      <BrowserRouter>
        <EpisodeDetail />
      </BrowserRouter>
    );
    const audio = container.querySelector('audio');
    expect(audio?.querySelector('source')?.getAttribute('src')).toBe(
      'https://verifi.podscribe.com/rss/p/traffic.libsyn.com/secure/jbpod/Joe_Budden_Podcast_778.mp3?dest-id=2422538'
    );
  });
});
