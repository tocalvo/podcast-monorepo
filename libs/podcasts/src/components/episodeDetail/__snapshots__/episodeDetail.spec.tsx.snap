// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EpisodeDetail > should render successfully 1`] = `
<div>
  <div
    class="_container_b2c8c3"
  >
    <div
      class="_card_33fd63"
      data-testid="podcasts_bar"
    >
      <a
        class="_image-link_33fd63"
        data-discover="true"
        href="/podcast/1535809341"
      >
        <img
          alt=""
          class="_image_33fd63"
          src="https://is1-ssl.mzstatic.com/image/thumb/Podcasts113/v4/f2/21/fa/f221fabd-017f-5125-633b-f1fe4f39802a/mza_182995249085044287.jpg/55x55bb.png"
        />
      </a>
      <hr
        class="_line_33fd63"
      />
      <a
        data-discover="true"
        href="/podcast/1535809341"
      >
        <p
          class="_title_33fd63"
        >
          The Joe Budden Podcast
        </p>
      </a>
      <a
        data-discover="true"
        href="/podcast/1535809341"
      >
        <p
          class="_italic_33fd63"
        >
          by 
          The Joe Budden Network
        </p>
      </a>
      <hr
        class="_line_33fd63"
      />
      <p
        class="_title2_33fd63"
      >
        Description:
      </p>
      <p
        class="_italic_33fd63"
      >
        Tune into Joe Budden and his friends. Follow along the crazy adventures of these very random friends.
      </p>
    </div>
    <div
      class="_card_b2c8c3 _episode-container_b2c8c3"
    >
      <h1>
        <b>
          Episode 778 | "Bottom Of The 9th"
        </b>
      </h1>
      <p>
        <i
          data-testid="episode_content"
        />
      </p>
      <audio
        class="_audio_b2c8c3"
        controls=""
      >
        <source
          src="https://verifi.podscribe.com/rss/p/traffic.libsyn.com/secure/jbpod/Joe_Budden_Podcast_778.mp3?dest-id=2422538"
          type="audio/mp3"
        />
        Your browser does not support the audio element.
      </audio>
    </div>
  </div>
</div>
`;
